### DISCUSSION TITLE
`on-monitor-detected` callback

### DISCUSSION THREAD

**Discussion by derrik-fleming** [nikitabobko/AeroSpace#958] [feature-ideas]
Not sure how feasible this is as a feature, but for those who want to execute a callback when a secondary display is detected a callback or condition for the existing callbacks could be useful. These are just two things that came to mind immediately when trying to solve this. I'm open to suggestions if this is a solved problem using the provided configuration, or otherwise. I think it'd be easily achievable using the CLI commands in a shell script but I'd appreciate a way for handling it within the config if possible.

I like the idea of binding applications to workspaces with mnemonic names, my browser lives on workspace `b`, terminal emulator on `t`, etc. I set up the window detection callbacks for the apps that I care about to move them to the desired workspaces. This works great when I don't have my external monitor connected. However, my main external monitor is a ultra-wide and more than capable of having 2 or 3 applications side-by-side. 

I want to bind applications to workspaces and I want separate bindings depending on whether or not specific external monitor(s) is connected. I also want to be able to execute commands when I connect my external monitor.

**Comment by fcaraujo** [comment]
++

I was about to suggest something similar and found this proposal. : )

I'm trying to run a command when the list of monitors changes (connecting/disconnection my 2nd monitor); in my case I want to trigger an event on my sketchybar as an example.

We could call it `exec-on-monitor-change` to follow the patter of the existing `exec-on-workspace-change`.

@nikitabobko I'm happy to jump on and learn how to contribute to the project if you or other maintainers are keen on having this in place, I'd just need some directions.

**Comment by nikitabobko** [comment]
The issue doesn't mention use cases

> I'm trying to run a command when the list of monitors changes (connecting/disconnection my 2nd monitor); in my case I want to trigger an event on my sketchybar as an example.

Why do you add AeroSpace in the mix? Shouldn't it be handled purely by sketchybar?

> We could call it exec-on-monitor-change to follow the patter of the existing exec-on-workspace-change.

I don't like `exec-on-workspace-change` callback, it's inconsistent with everything else. I hope that eventually `exec-on-workspace-change` will be replaced with: `on-workspace-change = 'exec-and-forget ...'`. But design questions of #278 and #351 has to be figured first

**Reply by Masstronaut** [reply]
Here's a concrete use case:
- When a display is disconnected, aerospace automatically moves all workspaces from the disconnected display to a connected display
- Sketchybar doesn't get notified this happens, and so incorrectly tries to render an item corresponding to that workspace on a display that the workspace is no longer associated with

If aerospace doesn't emit an event when this happens, I'm unable to gracefully handle it in sketchybar (or whatever other software I might be using that needs to respond to it).

When I manually move a workspace to another display using an aerospace keybind, I have the option to explicitly run other commands to handle it; [that is what I do](https://github.com/Masstronaut/dotfiles/commit/8cb79db1101968d70d848700e438d0026438b72e#diff-fd9c500732574eb553dd1290aae7073ebb68f1e026ff04e9085e0ea841ce24d1R204):

```sh
alt-shift-tab = ['move-workspace-to-monitor --wrap-around next', 'exec-and-forget sketchybar --trigger change-workspace-monitor TARGET_MONITOR=$(aerospace list-monitors --focused --format "%{monitor-appkit-nsscreen-screens-id}") TARGET_WORKSPACE=$(aerospace list-workspaces --focused)']
```

When the workspace are moved automatically by aerospace because a display is disconnected, I don't have the option to do this. The same applies in reverse.

**Comment by derrik-fleming** [comment]
> The issue doesn't mention use cases

@nikitabobko I've edited the original comment to include more details about my specific desires and use case.

**Comment by stschaef** [comment]
Building off of the original feature request

I'm looking for functionality that would allow aerospace to assign workspaces to take display precedence when an external monitor is connected. For instance, I'd like display my `b`, `m`, and `c` on each display by default when I connect to my external monitors at work, but aerospace will display workspaces `1`, `2`, and `3`. A display callback could let me focus these upon connection

I also have a different monitor at home, and I'd like to configure some default tiling to trigger when that display is connected.

**Comment by fcaraujo** [comment]
@nikitabobko trying to oversimplify the use case(s): it's achieving the same or similar things you'd be able to do through `exec-on-workspace-change`.
I want to trigger X when I switch workspace => I want to trigger Y when I connect/disconnect a different monitor.

Integration with Sketchybar was just a silly example. : )
However, thinking about it, I could expand: I can have a single item on my bar, and when I switch workspaces it changes its label to the workspace id; let's call it "A" and "B".
Now let's say I have "B" automatically assigned to go into my secondary monitor.
So when I connect the 2nd monitor, I want to change the label item to "A | X" (and IMHO Sketchybar has no responsibility to know that, same as switching workspaces basically).

I know naming things is polemical, and it was a dummy suggestion; I'm not too fussed about it and just really interested in the capability itself (by the way, personaly I like and follow `on-something` when I'm developing - similarly to what you already stated anyway).

Keep up the great work mate and thanks for your attention.

**Comment by timharris777** [comment]
I was looking for this as well and came across this feature request. +1

It would be extremely useful for me to be able to exec a command when aerospace detects that a new monitor has been added or removed. Yabai currently offers this with the following:

```
yabai -m signal --add event=display_added action="command you want to run"
yabai -m signal --add event=display_removed action="command you want to run"
```

**Comment by MediosZ** [comment]
I’m using AeroSpace and Sketchybar, and I’d love to get a notification when I connect or disconnect external monitors. That way, I can easily rearrange my Sketchybar settings. 

I agree this should be done by sketchybar purely for this single use case.

I come up with a simple solution for sketchybar:
```lua
local display_info = sbar.query("displays")

local display_detector = sbar.add("item", "display_detector", {
    width = 0,
    drawing = true
})

display_detector:subscribe("display_change", function(_)
    local info = sbar.query("displays")
    if not equals(display_info, info, true) then
        sbar.exec("sketchybar --reload")
        display_info = info
    end
end)
```

I think this can be super helpful for those who have similar use cases.

**Comment by alexcamlo** [comment]
Another use case:
In office I use a 38" monitor with layout:
```
h_tiles
    v_accordion
    v_accordion
````
But when I'm on the move and just using the laptop's screen I want to have:
```
v_accordion
```

So it would be cool to have a way of detecting the monitor and changing the layout accordinly, or even just flattening the workspace (until we have the layout saving functionality)

**Reply by zhuoqun-chen** [reply]
> Another use case: In office I use a 38" monitor with layout:
> 
> ```
> h_tiles
>     v_accordion
>     v_accordion
> ```
> 
> But when I'm on the move and just using the laptop's screen I want to have:
> 
> ```
> v_accordion
> ```
> 
> So it would be cool to have a way of detecting the monitor and changing the layout accordinly, or even just flattening the workspace (until we have the layout saving functionality)

I'd like this as well!

**Reply by jussapaavo** [reply]
I'm looking for this kind of function as well.

Instead of `on-monitor-detected` callback, maybe it could be:
```
[on-window-detected]
  if.monitor = 'built-in'
```
So that the window layout is changed to root h_accordion if there is only built-in, and instead if it's main and not built-in it's h_tiles with v_accordion.

**Comment by BenjaminBini** [comment]
Same here, I think there are a lot of legitimate use cases. Wanting the configuration of our windows to be different on spaces depending on monitor connected does not seem out  of scope of Aerospace.
